### DISCUSSION TITLE
`on_metadata` and `on_track` do not return a valid source anymore

### DISCUSSION THREAD

**Discussion by add-n2x** [savonet/liquidsoap#2509] [User Support]
**Describe the bug**
The [documentation on migrating to 2.1.x](https://www.liquidsoap.info/doc-dev/migrating.html) states, that `on_track` and `on_metadata` now return sources again, as previously in the 1.4.x releases:

> source.on_metadata and source.on_track now return a source as this was the case in previous versions,
> and associated handlers are triggered only when the returned source is pulled

But it seems the return value is not a complete `source` object. E.g. missing the `id` property.

**To Reproduce**

```py
url = "https://securestream.o94.at/live.mp3"
input_stream = input.http(id="input_stream", url)
def handle_meta(meta) =
    print("Called metadata handler and got: #{meta}")
end

# Works
id = input_stream.id()
print("Source ID: #{id}")

# Causes error
input_stream = input_stream.on_metadata(id="input_stream", handle_meta)
id = input_stream.id()
print("Source ID: #{id}")
```
The resulting error when trying to execute the script is:

> Error 6: Cannot apply that parameter because the function at test_source.liq, line 41, char 15-40
> has no argument labeled "id"!

Likewise `source.id(input_stream)` doesn't work either.

**Expected behavior**

No error with the code above.

**Version details**
- Liquidsoap latest 2.1.x rolling release

**Install method**
Debian package amd64

**Comment by toots** [comment]
Thanks for your feedback. This is currently the way it is intended. You have:
* `source.on_metadata` which is an operator (a function) that takes a source and returns a new source:
```
% liquidsoap -h source.on_metadata

Call a given handler on metadata packets.

Type: (?id : string?, source(audio='a, video='b, midi='c),
 (([string * string]) -> unit)) -> source(audio='a,
video='b, midi='c)
```
* if `foo` is a source _variable_, then it has. method `on_metadata`. This method returns nothing.

Your code can be written two different ways:
```ruby
url = "https://securestream.o94.at/live.mp3"
input_stream = input.http(id="input_stream", url)
def handle_meta(meta) =
    print("Called metadata handler and got: #{meta}")
end

# Works
id = input_stream.id()
print("Source ID: #{id}")

input_stream.on_metadata(handle_meta)
id = input_stream.id()
print("Source ID: #{id}")
```
Or:
```ruby
url = "https://securestream.o94.at/live.mp3"
input_stream = input.http(id="input_stream", url)
def handle_meta(meta) =
    print("Called metadata handler and got: #{meta}")
end

# Works
id = input_stream.id()
print("Source ID: #{id}")

input_stream = source.on_metadata(id="input_stream", url = "https://securestream.o94.at/live.mp3"
input_stream = input.http(id="input_stream", url)
def handle_meta(meta) =
    print("Called metadata handler and got: #{meta}")
end

# Works
id = input_stream.id()
print("Source ID: #{id}")

input_stream = source.on_metadata(id="input_stream", input_stream, handle_meta)
id = input_stream.id()
print("Source ID: #{id}")handle_meta)
id = input_stream.id()
print("Source ID: #{id}")
```
@smimram did that changes in https://github.com/savonet/liquidsoap/pull/2103 and this had to do with making sure that handlers are only called when the returned source is being used while handlers registered directly using the source method will be called whenever the source is being consumed. This is mostly an internal concern to us, I believe but maybe he might want to add more to this discussion.

**Reply by add-n2x** [reply]
Thanks for the clarification.

Good to know that, that this:

```py
input_stream.on_metadata(handle_meta)
```
and that is working:

```py
input_stream = source.on_metadata(id="input_stream", input_stream, handle_meta)
```
> this had to do with making sure that handlers are only called when the returned source is being used while handlers registered directly using the source method will be called whenever the source is being consumed. This is mostly an internal concern to us...

That's interesting. If "_being used_" means, that it affects the behavior of enabled/disabled/silenced sources, I gladly would love to know more. Since I may come up with similar questions soon. But since you write it is mostly an internal concern...no worries.

Many thanks for your help again.

**Reply by add-n2x** [reply]
Oh, so now it gets really interesting. I have done following test case:

```py
url = "https://orf-live.ors-shoutcast.at/oe1-q2a"
input_stream = input.http(id="input_stream", url, start=true)
def handle_meta1(meta) =
    print("Called metadata handler1 and got: #{meta}")
end
def handle_meta2(meta) =
    print("Called metadata handler2 and got: #{meta}")
end

# Variant A.) Attach `on_metadata` handler directly: handlers registered
# directly using the source method will be called whenever the source is being consumed
input_stream.on_metadata(handle_meta1)

# Variant B.) Source returned by `on_metadata`: Handlers are only called
# when the returned source is being used
input_stream = source.on_metadata(id="input_stream", input_stream, handle_meta2)
```

> Called metadata handler1 and got: ...
> Called metadata handler1 and got: ...

If the `output` is added, both handlers are called:

```py
output.alsa(device="default", mksafe(input_stream), bufferize=true)
```

> Called metadata handler1 and got: ...
> Called metadata handler1 and got: ...
> Called metadata **handler2** and got: ...

The same is true for a mixer i.e. when the source is selected and made audible via telnet:

```py
mixer = mix(id="mixer", [input_stream])
output.alsa(device="default", mksafe(mixer), bufferize=true)
```
> Called metadata handler1 and got: ...
> Called metadata handler1 and got: ...
> Called metadata **handler2** and got: ...

Glad the initial question brought me to a complety different area. This background info was really good to know. Merci.

**Reply by toots** [reply]
For sure. This multiple call situation is the reason @smimram brought back the old `source.on_metadata` operator.

**Comment by add-n2x** [comment]
@smimram there is unfortunately another case which doesn't work anymore:

```py
url = "https://orf-live.ors-shoutcast.at/oe1-q2a"
in_stream = input.http(id="in_stream", start=false, url)
in_stream = insert_metadata(id="in_stream", in_stream)
in_stream = source.on_metadata(id="in_stream", in_stream, on_metadata_notification)
```

Then calling this results in an error:

```py
in_stream.insert_metadata([("foo", "bar")])
```

>        **in_stream**.insert_metadata(metadata)
>
> Error 5: this value has no method `insert_metadata`

A workaround might be to refer to the initial input source with a different var and ID only. But is this intended, that a chain of wrapped sources is not addressable anymore?

---
**Edit**: I also notice in the logs, that switching between sources doesn't reference the actual sources anymore, but some auto-generated ID of the metadata source:

> 2022/07/14 12:51:17 [switch_2:3] Switch to on_metadata_2.

**Reply by toots** [reply]
This is intended. The `insert_metadata` method is only available on the source returned by the `insert_metadata` operator. Once it's passed to the `source.on_metadata`, the resulting source does not have that method anymore. You need to do:
```ruby
url = "https://orf-live.ors-shoutcast.at/oe1-q2a"
in_stream = input.http(id="in_stream", start=false, url)
in_stream = insert_metadata(id="in_stream", in_stream)

# Save the callback
insert_metadata = in_stream.insert_metadata

# Continue
in_stream = source.on_metadata(id="in_stream", in_stream, on_metadata_notification)

# Later:
insert_metadata(...)
```

**Reply by add-n2x** [reply]
Thanks for the clarification!
